qiskit/synthesis/mcx_synthesis.py

```diff
@@ -0,0 +1,273 @@
+        from qiskit.circuit.quantumcircuit import QuantumCircuit
+        num_qubits = num_ctrl_qubits + self.get_num_ancilla_qubits(num_ctrl_qubits) + 1
+
+        q = QuantumRegister(num_qubits, name="q")
+        qc = QuantumCircuit(q, name=self.name)
+        if num_qubits == 4:
+            qc._append(C3XGate(), q[:], [])
+        elif num_qubits == 5:
+            qc._append(C4XGate(), q[:], [])
+        else:
+            for instr, qargs, cargs in self._recurse(q[:-1], q_ancilla=q[-1]):
+                qc._append(instr, qargs, cargs)
+        return qc
+
+    def _recurse(self, q, q_ancilla=None):
+        # recursion stop
+        if len(q) == 4:
+            return [(C3XGate(), q[:], [])]
+        if len(q) == 5:
+            return [(C4XGate(), q[:], [])]
+        if len(q) < 4:
+            raise AttributeError("Something went wrong in the recursion, have less than 4 qubits.")
+
+        # recurse
+        num_ctrl_qubits = len(q) - 1
+        middle = ceil(num_ctrl_qubits / 2)
+        first_half = [*q[:middle], q_ancilla]
+        second_half = [*q[middle:num_ctrl_qubits], q_ancilla, q[num_ctrl_qubits]]
+
+        rule = []
+        rule += self._recurse(first_half, q_ancilla=q[middle])
+        rule += self._recurse(second_half, q_ancilla=q[middle - 1])
+        rule += self._recurse(first_half, q_ancilla=q[middle])
+        rule += self._recurse(second_half, q_ancilla=q[middle - 1])
+
+        return rule
+
+class MCXSynthesisVChain(MCXSynthesis):
+    """Implement the multi-controlled X gate using a V-chain of CX gates."""
+
+    def __init__(self, dirty_ancillas):
+        self.name = "mcx_vchain"
+        self.dirty_ancillas = dirty_ancillas
+
+    @staticmethod
+    def get_num_ancilla_qubits(num_ctrl_qubits):
+        return max(0, num_ctrl_qubits - 2)
+
+    def synthesize(self, num_ctrl_qubits):
+        # pylint: disable=cyclic-import
+        from qiskit.circuit.quantumcircuit import QuantumCircuit
+
+        num_qubits = num_ctrl_qubits + self.get_num_ancilla_qubits(num_ctrl_qubits) + 1
+
+        q = QuantumRegister(num_qubits, name="q")
+        qc = QuantumCircuit(q, name=self.name)
+        q_controls = q[: num_ctrl_qubits]
+        q_target = q[num_ctrl_qubits]
+        q_ancillas = q[num_ctrl_qubits + 1 :]
+
+        definition = []
+
+        if self.dirty_ancillas:
+            i = num_ctrl_qubits - 3
+            ancilla_pre_rule = [
+                (U2Gate(0, numpy.pi), [q_target], []),
+                (CXGate(), [q_target, q_ancillas[i]], []),
+                (U1Gate(-numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
+                (U1Gate(numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_target, q_ancillas[i]], []),
+                (U1Gate(-numpy.pi / 4), [q_ancillas[i]], []),
+                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
+                (U1Gate(numpy.pi / 4), [q_ancillas[i]], []),
```

Comment:
Could we replace the U1/2 gates with the standard gates we use now? (Same holds for the definitions below)
```suggestion
                (HGate(), [q_target], []),
                (CXGate(), [q_target, q_ancillas[i]], []),
                (PhaseGate(-numpy.pi / 4), [q_ancillas[i]], []),
                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
                (PhaseGate(numpy.pi / 4), [q_ancillas[i]], []),
                (CXGate(), [q_target, q_ancillas[i]], []),
                (PhaseGate(-numpy.pi / 4), [q_ancillas[i]], []),
                (CXGate(), [q_controls[-1], q_ancillas[i]], []),
                (PhaseGate(numpy.pi / 4), [q_ancillas[i]], []),
```